app/controllers/UpdateRelationshipController.scala

```diff
@@ -19,291 +19,261 @@ package controllers
+      (updateRelationshipService.getRelationshipRecords map { relationshipRecords =>
+        Ok(views.html.coc.bereavement(relationshipRecords.primaryRecord.role))
+      }) recover handleError
+  }
+
+  def divorceEnterYear: Action[AnyContent] = authenticate.async {
+    implicit request =>
+      updateRelationshipService.getDivorceDate map { optionalDivorceDate =>
+        optionalDivorceDate.fold(Ok(views.html.coc.divorce_select_year(DivorceSelectYearForm.form))){ divorceDate =>
+          Ok(views.html.coc.divorce_select_year(DivorceSelectYearForm.form.fill(divorceDate)))
+        }
+      } recover {
+        case NonFatal(_) =>
+          Ok(views.html.coc.divorce_select_year(DivorceSelectYearForm.form))
       }
   }
 
-  private[controllers] def getConfirmationInfoFromReason(reason: EndRelationshipReason,
-                                                         cacheData: UpdateRelationshipCacheData): (Boolean, Option[LocalDate], Option[LocalDate]) = {
-    reason.endReason match {
-      case EndReasonCode.DIVORCE_PY =>
-        (false, None, Some(timeService.getEffectiveDate(reason)))
+  def submitDivorceEnterYear: Action[AnyContent] = authenticate.async {
+    implicit request =>
+      DivorceSelectYearForm.form.bindFromRequest.fold(
+        formWithErrors => {
+          Future.successful(BadRequest(views.html.coc.divorce_select_year(formWithErrors)))
+        }, {
+          case divorceDate =>
+            updateRelationshipService.saveDivorceDate(divorceDate) map { _ =>
+              Redirect(controllers.routes.UpdateRelationshipController.divorceEndExplanation())
+          }
+        }
+      ) recover handleError
+  }
 
-      case EndReasonCode.DIVORCE_CY =>
-        (false, None, timeService.getEffectiveUntilDate(reason))
+  def divorceEndExplanation: Action[AnyContent] = authenticate.async {
+    implicit request =>
+      (for {
+        (role, divorceDate) <- updateRelationshipService.getDataForDivorceExplanation
+        datesForDivorce = updateRelationshipService.getMAEndingDatesForDivorce(role, divorceDate)
+        _ <- updateRelationshipService.saveMarriageAllowanceEndingDates(datesForDivorce)
+      } yield {
+       val viewModel = DivorceEndExplanationViewModel(role, divorceDate, datesForDivorce)
+        Ok(views.html.coc.divorce_end_explanation(viewModel))
+      }) recover handleError
+  }
 
-      case EndReasonCode.CANCEL =>
-        (false, None, timeService.getEffectiveUntilDate(reason))
 
-      case EndReasonCode.REJECT =>
-        val selectedRelationship = updateRelationshipService.getRelationship(cacheData)
-        val isEnded = selectedRelationship.participant1EndDate.exists(_ != "")
+  def confirmEmail: Action[AnyContent] = authenticate.async {
+    implicit request =>
+      lazy val emptyEmailView = views.html.coc.email(emailForm)
+      updateRelationshipService.getEmailAddress map {
```

Comment:
could this service return and Option[EmailAddress] rather that Option[String] ?

Reply:
Fixed.